proposals/0000-componentized-cabal.rst

```diff
@@ -0,0 +1,360 @@
+the possible inputs into the build, with the hope that the hash uniquely
+determines all of the build products of the package), and
+``--dependency=pkgname=ipid-1.0-abcdef`` specifies a specific dependency
+to be used during the build of all the components.
+``--exact-configuration`` requires ALL dependencies of all enabled
+components to be specified explicitly.  Notice that this particular
+command has ``--enable-tests`` passed (thus enabling the building of the
+test suite) and specifies which version of ``tasty`` (a test library)
+should be used. This command configures both the library and the test
+suite of this package *simultaneously*.
+
+Currently, the ``configure`` command takes no arguments.
+
+**The proposal.** We extend ``configure`` to take a single argument
+specifying which component is to be configured::
+
+    ./Setup configure test-suite
+
+It is invalid to pass multiple arguments. Under this mode, the configure
+command behaves differently in the following ways:
+
+1. Subsequent commands, e.g., build, *no longer* use an inplace package
+   database (conventionally in ``dist/package.conf.inplace``). It is
+   expected that all dependencies live in the database stack specified
+   by the user.
+
+2. The set of required ``--dependency`` flags for an
+   ``--exact-configuration`` is changed to precisely the ``build-depends``
+   of the component being enabled. Dependencies from non-enabled
+   components do not have to be specified (similar to how
+   ``--disable-tests`` works today). However, an *internal* dependency,
+   e.g., from a test suite to the library, must now be specified as
+   ``--dependency=packagename=pkgipid-1.0-abcdefg123``. In the presence of
+   convenience libraries, there may be multiple such extra "internal"
+   dependencies. These are disambiguated from ordinary dependencies by
+   inspection of the package description.
+
+3. We replace the ``--ipid`` flag with a ``--cid`` flag, which specifies the
+   *component identifier* of the component being built. The user must
+   pick a distinct ``--cid`` for each component in the package, as
+   distinct from an installed package identifier which was global for a
+   package.  This flag is only valid when a component is explicitly
+   specified to be built. For backwards compatibility, the ``--ipid`` flag
+   retains its original meaning as an identifier for the package as a
+   whole: thus when we build a component with it, the IPID is qualified
+   with the name of the component to form a component id.
+
+
+4. If a library contains a ``build-tools`` dependency on an executable, if
+   the library is being configured by itself, it is expected that the
+   caller arrange for the executable to be present on the ``PATH``, so
+   that the build tool configuration process succeeds. (This currently
+   happens automatically for internal executables.)
+
+5. While a user can still explicitly specify installation paths for
```

Comment:
Well, my thinking was that each component should get its own "install path" of sorts. But now that I think about it maybe it is not so important that Cabal sets this up; any package manager that care (e.g., cabal-install) can set the paths appropriately.